File: Week3/1. Reverse_Words_in_a_String.py
"""
Using Deque

Complexity Analysis

Time complexity: O(N)

Space complexity: O(N)

"""

from collections import deque

class Solution:
    def reverseWords(self, s: str) -> str:
        #return ' '.join(s.split()[::-1])
        
        #implemetation for deque:
        
        left = 0
        right = len(s) - 1
        
        #removing the leading spaces;
        while left <= right and s[left] == ' ':
            left = left + 1
        
        #removing trailing spaces:
        while left <= right and s[right] == ' ':
            right = right - 1
            
        d, word = deque(), []
        
        #pushing word by word into deque
        while left <= right:
            if s[left] == ' ' and word:
                d.appendleft(''.join(word))
                word = []
            elif s[left] != ' ':
                word.append(s[left])
                
            left += 1
            
        d.appendleft(''.join(word))
        
        return ' '.join(d)
                
            
        
        
        

File: Week3/1. Reverse_Words_in_a_String.cpp
/*

Given an input string, reverse the string word by word.

 

Example 1:

Input: "the sky is blue"
Output: "blue is sky the"
Example 2:

Input: "  hello world!  "
Output: "world! hello"
Explanation: Your reversed string should not contain leading or trailing spaces.
Example 3:

Input: "a good   example"
Output: "example good a"
Explanation: You need to reduce multiple spaces between two words to a single space in the reversed string.
 

Note:

A word is defined as a sequence of non-space characters.
Input string may contain leading or trailing spaces. However, your reversed string should not contain leading or trailing spaces.
You need to reduce multiple spaces between two words to a single space in the reversed string.
 

Follow up:

For C programmers, try to solve it in-place in O(1) extra space.
*/

class Solution {
public:
    string reverseword(string &s, int i, int j){
        while(i<j){
          char t=s[i];
          s[i++]=s[j];
          s[j--]=t;
        } 
        return s;
    }
    
    
    string reverseWords(string s) {
        int i=0, j=0;
        int l=0;
        int len=s.length();
        int wordcount=0;
        
        while(true){
            while(i<len && s[i] == ' ') i++;  // skip spaces in front of the word
            if(i==len) break;
            if(wordcount) s[j++]=' ';
            l=j;
            while(i<len && s[i] != ' ') {s[j]=s[i]; j++; i++;} 
            reverseword(s,l,j-1);                // reverse word in place
            wordcount++;
            
        }
        
        s.resize(j);                           // resize result string
        return reverseword(s,0,j-1);                  // reverse whole string
    
        
    }
};
